### PR TITLE
Refactor/config and prometheus metrics

### LINKED ISSUE #572: Expose message processing statistics to prometheus
Would be nice to be able to alert on the processing metrics times.

This should be possible via a promtheus metrics exporter or a /metrics endpoint so prometheus can scrape it.

Table example:
```
┌─────────┬───────────────────┬───────┬─────────────┬─────────┬────────┬────────┐
│ (index) │     eventType     │ count │ avgCountSec │ avgTime │  min   │  max   │
├─────────┼───────────────────┼───────┼─────────────┼─────────┼────────┼────────┤
│    0    │      'Death'      │  275  │   '4.58'    │ '0.63'  │ '0.11' │ '1.21' │
│    1    │ 'FacilityControl' │   4   │   '0.07'    │ '0.67'  │ '0.52' │ '1.00' │
│    2    │ 'GainExperience'  │ 2833  │   '47.22'   │ '0.64'  │ '0.09' │ '1.21' │
│    3    │ 'VehicleDestroy'  │  63   │   '1.05'    │ '0.65'  │ '0.13' │ '1.21' │
└─────────┴───────────────────┴───────┴─────────────┴─────────┴────────┴────────┘
```

TODO:
- [x] Refactor all `ps2AlertsApiClient` calls into `PS2AlertsApiDriver`
- [x] Refactor all `Rest.Census` query builders into `CensusRequestDriver`
- [ ] Apply all metrics
- [ ] Double check all metric points
- [ ] Remove old `logMetrics` calls

Should be able to add these as data points within the `TimingStatisticsAuthority` and pull it out.

# Metrics to collect

## Gauges
- [x] InstancesActive - number of alerts currently running
- [ ] Messages in flight - number of messages being processed

### RabbitMQ messages
- [x] Totals
- [x] Success
- [x] Retried
- [x] Delayed (not implemented yet)
- [x] Discarded (failures)

### Events (success & failures for each):
- [ ] Death
  - [x] Count
  - [ ] Success
  - [ ] Fail
- [ ] FacilityControl
  - [x] Count
  - [ ] Success
  - [ ] Fail
- [ ] GainExperience
  - [x] Count
  - [ ] Success
  - [ ] Fail
- [ ] VehicleDestroy
  - [x] Count
  - [ ] Success
  - [ ] Fail
- [ ] MetagameEvent
  - [x] Count
  - [ ] Success
  - [ ] Fail

### Instances
- [x]  Successfully Started
- [x]  Failed to start
 
#### Census
- [x] /character
- [x] /item
- [x] /map_region
#### Falcon
- [x]  /item
#### PS2Alerts API
- [x] /censusRegions
- [x] /instances 

### Cache
Hits and misses by:
- [x] Character
- [x] Item
- [x] FacilityData
- [x] MapRegion
 
### Processing time
- [x] Death
- [x] FacilityControl
- [x] GainExperience
- [x] VehicleDestroy
- [ ] MetagameEvents

#### Census
- [x] /character
- [x] /item
- [x] /map_region
#### Falcon
- [x] /item
#### PS2Alerts API
- [x] /censusRegions
- [x] /instances

### PR DESCRIPTION
**Pull request by microwavekonijn**
Refactor that implements the nestjs config module, and added prom-client so we can expose metrics to prometheus. The current metrics only consist of nodejs metrics only, but can easily be expanded.

❗ This PR is used as a basis for another PR, so please hold off with merging and changes until it is decided to either accept or reject the downstream PR.

Closes #572.

### CODE REVIEW

**Review by Maelstromeous — APPROVED** [review]

File: src/modules/rabbitmq/factories/RabbitMQQueueFactory.ts [hunk 1]
```diff
@@ -30,6 +33,7 @@ export default class RabbitMQQueueFactory {
         return new InstanceEventQueue(
             this.connectionManager,
             queueName,
+            this.config.get('rabbitmq.topicExchange'),
```
Comment (Maelstromeous) [inline src/modules/rabbitmq/factories/RabbitMQQueueFactory.ts:36]: Not quite understanding why we've abstracted the exchange into an argument, this caused a bug where this particular line was the wrong exchange type, which wouldn't have been possible previously as we were pulling from the config object.

Could you explain why here we're creating an construction injection yet they're different queue types, and in other places we're not?

**Review by Maelstromeous — CHANGES_REQUESTED** [review]
Please see concerns raised

**Comment by Maelstromeous** [comment]
Will close #572.

File: src/main.ts [hunk 2]
```diff
@@ -1,13 +1,11 @@
...
-import config from './config';
 import {FastifyAdapter} from '@nestjs/platform-fastify';
 import {Logger} from '@nestjs/common';
+import {ConfigService} from '@nestjs/config';
 
 async function bootstrap(): Promise<void> {
-    process.on('unhandledRejection', (err) => {
-        throw err;
-    }).on('uncaughtException', (err) => {
+    process.on('uncaughtException', (err) => {
```
[the remark this answers is not in the document]
Comment (microwavekonijn) [reply src/main.ts:8]: With Node 18 rejection can be caught as as uncaught exceptions. No need for more work

**Comment by Maelstromeous** [comment]
Closes #572

File: src/modules/rabbitmq/factories/RabbitMQQueueFactory.ts (hunk 1 above)
Comment (microwavekonijn) [reply src/modules/rabbitmq/factories/RabbitMQQueueFactory.ts:36]: Before we accessed a property, now we fetch a config value. This seemed more appropriate, as there is more overhead now.

Comment (Maelstromeous) [reply src/modules/rabbitmq/factories/RabbitMQQueueFactory.ts:36]: Sure, but couldn't we pull that config item out of the config in the class directly rather than being passed in?

Comment (microwavekonijn) [reply src/modules/rabbitmq/factories/RabbitMQQueueFactory.ts:36]: Why should the `InstanceEventQueue` have a dependency on the `ConfigService`? It just needs a name nothing more. Limiting the spread of dependencies will be better in the long term.

File: src/handlers/ps2census/MetagameEventEventHandler.ts [hunk 3]
```diff
@@ -24,6 +23,8 @@ import {ps2AlertsApiEndpoints} from '../../ps2alerts-constants/ps2AlertsApiEndpo
 export default class MetagameEventEventHandler implements QueueMessageHandlerInterface<MetagameEvent> {
     private static readonly logger = new Logger('MetagameEventEventHandler');
 
+    private readonly metagameCreationsEnabled = true;
```
[the remark this answers is not in the document]
Comment (microwavekonijn) [reply src/handlers/ps2census/MetagameEventEventHandler.ts:26]: Ah, yeah. I mostly did it as the value was hard coded in the config.

File: src/authorities/MetricsAuthority.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+        await Promise.all(promises).then((results) => {
+            this.metricsHandler.setGauge(METRICS_NAMES.CACHE_GAUGE, results[0].length, {type: 'cache_character'});
+            this.metricsHandler.setGauge(METRICS_NAMES.CACHE_GAUGE, results[1].length, {type: 'cache_item'});
+            this.metricsHandler.setGauge(METRICS_NAMES.CACHE_GAUGE, results[2].length, {type: 'cache_facility_data'});
+            this.metricsHandler.setGauge(METRICS_NAMES.CACHE_GAUGE, results[3].length, {type: 'unknown_items'});
+            this.metricsHandler.setGauge(METRICS_NAMES.CACHE_GAUGE, results[4].length, {type: 'unknown_facilities'});
+            this.metricsHandler.setGauge(METRICS_NAMES.CACHE_GAUGE, results[5].length, {type: 'character_presence'});
+            this.metricsHandler.setGauge(METRICS_NAMES.CACHE_GAUGE, results[6].length, {type: 'outfit_participants'});
+        });
```
Comment (microwavekonijn) [inline src/authorities/MetricsAuthority.ts:80]: I think this is unnecessary. Defaults are not needed as Prometheus is able to infer this by itself. As for the Redis stuff maybe consider select? To the Gauge configuration add a collect method that runs a query on Redis. One of the considerations is that metrics are suppose to be simple addition, hence there shouldn't be a need for these kinds of complexities.

Example gauge provider:
```ts
{
  provider: getToken('tralala_total'),
  useFactory: (redis: Redis) => getOrCreateMetric('Gauge', {
    name: 'tralala_total',
    help: 'Lorem ipsum',
    collect: () =>{
      this.set(redis.someQueryThatReturnsNumberWrappedInPromise());
    }
},
inject: [Redis]

File: src/drivers/CensusApiRetryDriver.ts [hunk 5]
```diff
@@ -1,51 +1,72 @@
...
         private readonly caller: string,
-    ) {}
+        private readonly metricsHandler: MetricsHandler,
+        config: ConfigService,
+    ) {
+        this.timeoutTime = config.get('census.timeout'); // Maximum Census tolerance
+        this.endpoint = `/${this.query.collection}`;
+    }
 
     public async try(attempts = 0): Promise<Rest.CensusResponse<Rest.Format<T>> | undefined> {
```
Comment (microwavekonijn) [inline src/drivers/CensusApiRetryDriver.ts:29]: As we talked about, the return is not the format that the query returns, but rather the base format of a collection without the joins.

File: src/modules/metrics/MetricsConstants.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    BROKER_COUNT: 'aggregator_broker_count',
+    CACHE_COUNT: 'aggregator_cache_count',
+    ERRORS_COUNT: 'aggregator_errors_count',
+    EXTERNAL_REQUESTS_COUNT: 'aggregator_external_requests_count',
+    INSTANCES_COUNT: 'aggregator_instances_count',
+    QUEUE_MESSAGES_COUNT: 'aggregator_queue_messages_count',
+    ZONE_MESSAGE_COUNT: 'aggregator_zone_message_count',
+
+    // Gauges
+    CACHE_GAUGE: 'aggregator_cache_gauge',
```
Comment (microwavekonijn) [inline src/modules/metrics/MetricsConstants.ts:12]: The postfix should say something about the variable. Gauge as a postfix seems inaccurate. The whole variable btw should be self describing encapsulating all properties that the variable represents. This one for example should be `cache_current_items_count`.

Below you also have the postfix `_histogram` which is even more inaccurate as histograms add their own postfixes. I know it is only semantics, but for Prometheus it is important.

File: src/modules/metrics/MetricsModule.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        PROM_METRICS.EXTERNAL_REQUESTS_COUNT,
+        PROM_METRICS.QUEUE_MESSAGES_COUNT,
+        PROM_METRICS.INSTANCES_COUNT,
+        PROM_METRICS.ZONE_MESSAGE_COUNT,
+        // Gauges
+        PROM_METRICS.CACHE_GAUGE,
+        PROM_METRICS.INSTANCES_GAUGE,
+        // Histograms
+        PROM_METRICS.EXTERNAL_REQUESTS_HISTOGRAM,
+        PROM_METRICS.EVENT_PROCESSING_HISTOGRAM,
```
Comment (microwavekonijn) [inline src/modules/metrics/MetricsModule.ts:90]: Don't do this, leave the naming of tokens up to the lib, in the case that this ever changes it needs to be updated manually. Use the `getToken()` function.

**Review by microwavekonijn — COMMENTED** [review]
There are some things that needs addressing.

**Review by Maelstromeous — DISMISSED** [review]

File: src/modules/metrics/MetricsController.ts [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+import {Controller, Get, Res} from '@nestjs/common';
+import {Response} from 'express';
+
+@Controller()
+export class MetricsController extends PrometheusController {
+    @Get()
+    public async index(@Res() response: Response) {
+        return super.index(response);
+    }
+}
```
[the remark this answers is not in the document]
Comment (Maelstromeous) [reply src/modules/metrics/MetricsController.ts]: Yeah I was using this to debug, it's gone now 3574b6b4e975040c717eab32c1397d235758d714

File: package.json [hunk 9]
```diff
@@ -28,19 +28,23 @@
...
     "@nestjs/microservices": "^9.4.0",
     "@nestjs/platform-fastify": "^9.4.0",
     "@nestjs/terminus": "^9.1.1",
+    "@willsoto/nestjs-prometheus": "^5.1.0",
     "amqp-connection-manager": "^4.1.12",
     "amqplib": "^0.10.0",
     "axios": "^1.3.5",
     "date-fns": "^2.29.3",
-    "dotenv": "^16.0.3",
+    "express": "^4.18.2",
```
[the remark this answers is not in the document]
Comment (Maelstromeous) [reply package.json:41]: ac3fa5cde1ac2be52eb9772226cf57da70efe0d5

File: src/AppModule.ts [hunk 10]
```diff
@@ -188,9 +205,13 @@ import {Death} from 'ps2census';
...
     ],
 })
-export default class AppModule implements OnApplicationShutdown {
+export default class AppModule implements OnApplicationShutdown, OnApplicationBootstrap {
     private readonly logger = new Logger('App');
 
+    public onApplicationBootstrap(): void {
+        this.logger.log('Houston we are go for launch!');
+    }
+
```
[the remark this answers is not in the document]
Comment (Maelstromeous) [reply src/AppModule.ts]: 1fb8e3c55d57ab99247ec476ee08721496dbb0ef

File: src/modules/metrics/MetricsConstants.ts (hunk 6 above)
Comment (Maelstromeous) [reply src/modules/metrics/MetricsConstants.ts]: 691f5257777e15bc7ecceca2ebeedf2f10a87035 and e9aa20201e3ecf99abc8e6b5966aac1ef8a56c62 to better describe some odd ones

> Below you also have the postfix _histogram which is even more inaccurate as histograms add their own postfixes. I know it is only semantics, but for Prometheus it is important.

Not sure if I'm misunderstanding but it doesn't become `aggregator_external_requests_histogram_bucket_histogram`, it's `aggregator_external_requests_histogram_bucket` which is correct?

File: src/modules/metrics/MetricsModule.ts (hunk 7 above)
Comment (Maelstromeous) [reply src/modules/metrics/MetricsModule.ts]: 2539bd50321e7eb1a335d37c714d4860ad7fdd1b

File: src/authorities/MetricsAuthority.ts (hunk 4 above)
Comment (Maelstromeous) [reply src/authorities/MetricsAuthority.ts]: Removed `setDefault()` 2bb13070dc04d2795893f0154526ec3b8207b3b9

Comment (Maelstromeous) [reply src/authorities/MetricsAuthority.ts]: In terms of the collect method, can this be achieved using labels? Also not quite sure where this provider is meant to be injected etc

File: src/drivers/CensusApiRetryDriver.ts (hunk 5 above)
Comment (Maelstromeous) [reply src/drivers/CensusApiRetryDriver.ts:29]: Honestly, I don't know how to resolve this. I need more ELI5 unfortunately!

File: src/modules/metrics/MetricsConstants.ts (hunk 6 above)
Comment (microwavekonijn) [reply src/modules/metrics/MetricsConstants.ts]: The misconception starts because a counter usually ends with `_count`. The naming should be like `<namspace>_<description>_<metric_description>`. Metric description should be related to what it is sum, count, bytes, total, etc. For a histogram and also summary prometheus adds its own `_bucket`, `_sum`, `_count`. The generated prometheus description of the metric group will include that it is of the type histogram.

Example from default node metrics:
```
# HELP nodejs_gc_duration_seconds Garbage collection duration by kind, one of major, minor, incremental or weakcb.
# TYPE nodejs_gc_duration_seconds histogram
nodejs_gc_duration_seconds_bucket{le="0.001",kind="minor"} 287
nodejs_gc_duration_seconds_bucket{le="0.01",kind="minor"} 388
nodejs_gc_duration_seconds_bucket{le="0.1",kind="minor"} 388
nodejs_gc_duration_seconds_bucket{le="1",kind="minor"} 388
nodejs_gc_duration_seconds_bucket{le="2",kind="minor"} 388
nodejs_gc_duration_seconds_bucket{le="5",kind="minor"} 388
nodejs_gc_duration_seconds_bucket{le="+Inf",kind="minor"} 388
nodejs_gc_duration_seconds_sum{kind="minor"} 0.38074574875831596
nodejs_gc_duration_seconds_count{kind="minor"} 388
nodejs_gc_duration_seconds_bucket{le="0.001",kind="incremental"} 4
nodejs_gc_duration_seconds_bucket{le="0.01",kind="incremental"} 4
nodejs_gc_duration_seconds_bucket{le="0.1",kind="incremental"} 4
nodejs_gc_duration_seconds_bucket{le="1",kind="incremental"} 4
nodejs_gc_duration_seconds_bucket{le="2",kind="incremental"} 4
nodejs_gc_duration_seconds_bucket{le="5",kind="incremental"} 4
nodejs_gc_duration_seconds_bucket{le="+Inf",kind="incremental"} 4
nodejs_gc_duration_seconds_sum{kind="incremental"} 0.000738584041595459
nodejs_gc_duration_seconds_count{kind="incremental"} 4
nodejs_gc_duration_seconds_bucket{le="0.001",kind="major"} 0
nodejs_gc_duration_seconds_bucket{le="0.01",kind="major"} 2
nodejs_gc_duration_seconds_bucket{le="0.1",kind="major"} 2
nodejs_gc_duration_seconds_bucket{le="1",kind="major"} 2
nodejs_gc_duration_seconds_bucket{le="2",kind="major"} 2
nodejs_gc_duration_seconds_bucket{le="5",kind="major"} 2
nodejs_gc_duration_seconds_bucket{le="+Inf",kind="major"} 2
nodejs_gc_duration_seconds_sum{kind="major"} 0.005652751088142395
nodejs_gc_duration_seconds_count{kind="major"} 2
```

You can always break with these conventions, but other people will have given this a great deal of though plus people will know what to expect with these conventions.

File: src/authorities/MetricsAuthority.ts (hunk 4 above)
Comment (microwavekonijn) [reply src/authorities/MetricsAuthority.ts]: I am pretty sure that `this.set()` in collect will accept labels if that is what you mean. The provider needs to be registered in any module or its own. Wherever you do it you just import redis or any other modules that hold the providers that you infer the metrics from, then inject the providers and build the metric. The metric itself doesn't have to be exported or injected anywhere, whenever the metrics endpoint will be scraped the collect method will be called first and set the metrics.
